File: LC/211/211.java
class Node {
    public Character chr;
    public boolean isWord;
    public Map<Character, Node> next;
    
    public Node(Character chr, boolean isWord) {
        this.chr = chr;
        this.isWord = isWord;
        this.next = new HashMap<Character, Node>();
    }
}

class WordDictionary {
    private Node trie;

    public WordDictionary() {
        this.trie = new Node(null, false);
    }
    
    public void addWord(String word) {
        Node node = this.trie;
        for (int i = 0; i < word.length(); i++) {
            char chr = word.charAt(i);
            
            if (!node.next.containsKey(chr)) node.next.put(chr, new Node(chr, false)); 
            node = node.next.get(chr);
        }
        node.isWord = true;
    }
    
    public boolean search(String word) {
        Node node = this.trie;
        return dfs(word, 0, node);
    }
    
    private boolean dfs(String word, int i, Node node) {
        if (i >= word.length()) return node.isWord;
        
        char chr = word.charAt(i);
        if (chr == '.') {
            for (Node nextNode: node.next.values()) {
                if (dfs(word, i+1, nextNode)) return true;
            }
        }
        return node.next.containsKey(chr) && dfs(word, i+1, node.next.get(chr));
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: LC/211/211.ts
class Node {
    val: string;
    isWord: boolean;
    next: {[key: string]: Node};
    
    constructor(val: string) {
        this.val = val;
        this.isWord = false;
        this.next = {};
    }
}

class WordDictionary {
    trie: Node;
    
    constructor() {
        this.trie = new Node("");
    }

    addWord(word: string): void {
        let node: Node = this.trie;
        
        for (let i = 0; i < word.length; i++) {
            const char: string = word[i];
            if (!Object.hasOwnProperty.call(node.next, char)) {
                node.next[char] = new Node(char);
            }
            node = node.next[char];
        }
        node.isWord = true;
    }

    search(word: string): boolean {
        return this.searchFromNode(word, 0, this.trie);
    }

    searchFromNode(word: string, start: number, node: Node): boolean {
        for (let i = start; i < word.length; i++) {
            const char: string = word[i];
            if (char === '.') {
                return Object.values(node.next)
                    .some((nextNode) => this.searchFromNode(word, i+1, nextNode));
            }
            
            if (!Object.hasOwnProperty.call(node.next, char)) return false;
            node = node.next[char];
        }
        return node.isWord;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */
